### ISSUE TITLE
Create Pull Requests from UI (or Custom Command)

### ISSUE THREAD

**Issue by ghost** [ForkIssues/TrackerWin#577]
Bitbucket Server has a REST API for this, but doesn't seem that easy to be able to use it from the custom command page.

https://developer.atlassian.com/server/bitbucket/how-tos/command-line-rest/

Would be really nice if there were some more integration with Bitbucket.

**Comment by jerone** [comment]
I would love to see this also implemented for **Azure DevOps** (Services), which my team uses.

There is also an [REST API](https://docs.microsoft.com/en-us/rest/api/azure/devops/git/?view=azure-devops-rest-5.1) available: 
https://docs.microsoft.com/en-us/rest/api/azure/devops/git/pull%20requests/create?view=azure-devops-rest-5.1

The [examples](https://github.com/microsoft/azure-devops-dotnet-samples/tree/master/ClientLibrary/Samples/Git) are even in C#.

**Comment by josesimoes** [comment]
Love to see this for GitHub too. The API is there for sure.

**Comment by ghost** [comment]
Would be a good pay feature...

**Comment by dreadnaut** [comment]
In the meanwhile, it is possible to open a pull request on a repository without any API work, at least in GitHub and BitBucket, by visiting the correct URL.

I set up [this script](https://gist.github.com/dreadnaut/9a91d663245e8fc8684d32a199fdc2d8) as a custom action, and I run it via Fork's `Ctrl-P`.

**Comment by DanPristupov** [comment]
In Fork 1.44 we added the ability to create pull requests.

**Comment by jerone** [comment]
@DanPristupov Great feature!

Some feedback on PR with **GitHub** (will try to test DevOps next week):

My situation; I forked an repo to my own account on GitHub (normal behavior). My fork is called "origin" and I included an remote to the original repo, called "upstream". I created an local branch, made an commit and pushed it to "origin", because pushing to "upstream" is blocked as I'm not an owner/collaborator of the "upstream" repo.

* Local branch -> Create PR on "origin"
  * opens link to "upstream" repo
    * base "upstream" master, head "origin" local branch
      * expected: to open link on "origin" repo.
* Local branch -> Create PR on "origin"
  * opens link to "upstream" repo
    * base "upstream" master, head "upstream" local branch
      * expected: head "origin" local branch
* "origin" branch -> Create PR on "origin"
  * opens link to "upstream" repo
    * base "upstream" master, head "origin" local branch
      * expected: to open link on "origin" repo.
      * suggestion: maybe this should also include the option to open on "upstream" remote.

**Comment by jerone** [comment]
@DanPristupov I just tested **Azure DevOps** and it is **not** working.

I'm being redirected to:
```
https://dev.azure.com/{organization}/_git/{repository}/pullrequestcreate?sourceRef=release/{branch}
```

While I should be redirected to:
```
https://dev.azure.com/{organization}/{project}/_git/{repository}/pullrequestcreate?sourceRef=release/{branch}
```
*Notice the inclusion of the `{project}` parameter.*

The remote url can be as following:
```
https://dev.azure.com/{organization}/{project}/_git/{repository}
git@ssh.dev.azure.com:v3/{organization}/{project}/{repository}
https://{organization}@dev.azure.com/{organization}/{project}/_git/{repository}
```

The first one I was using to test this.

According [the specs](https://docs.microsoft.com/en-us/rest/api/azure/devops/git/pull%20requests/create?view=azure-devops-rest-5.1#uri-parameters), the parameter is optional, but probably required for organization with more then one project.

Related to https://github.com/ForkIssues/TrackerWin/issues/641.

**Comment by ghost** [comment]
Doesn't seem to work on Bitbucket Server, for which I created this ticket against.  Unless I am missing something.   Right clicking on a branch, I see nothing about creating pull requests.  Is there something I need to set up?

**Comment by MarchingCube** [comment]
I also wonder about support for Azure DevOps Server, is this something that can be supported? It is a on-premises version and I wonder how does it detect that it is dealing with Azure DevOps, since it is not detecting our Server instance.

**Comment by DanPristupov** [comment]
@jerone cross repository pull requests are not supported yet.

We didn't know about organizations in Azure, fixing this right now.

@asm-mgillespie private BitBucket instances can't be supported because they have different url. Their REST API is not an option, unfortunately. It's an unbelievable mess and I don't want to spend my time on that.

@MarchingCube this is similar to BitBucket Server. Custom urls are not supported. We will try to figure out a solution in the future.

**Comment by ghost** [comment]
You don't want to "bother" implementing Enterprise support, yet you want us to "bother" paying for the app?

**Comment by DanPristupov** [comment]
No, it's a bit different. We do our best implementing a product and you buy it if you like it.

We are not a big company with a huge team. I told multiple times that there's only Tanya and I in Fork and we spend our time very carefully. Currently we have more important tasks.

**Comment by markmonster** [comment]
> @DanPristupov I just tested **Azure DevOps** and it is **not** working.
> 

Additionally I also found that the "branche" name isn't url-escaped. This is also causing issues.

**Comment by DanPristupov** [comment]
Could you give an example of such branch name please?

**Comment by markmonster** [comment]
Yes, an example:

feature/#77928_F4_-_Klikservice_-_Klant_ontvangt_email_met_Prijsupdate_nadat_prijs_bekend_is

**Comment by DanPristupov** [comment]
Hopefully we have fixed all the problems with Azure DevOps in 1.45:
- support for organizations
- support for visualstudio.com
- support for branches with / and # symbols.

**Comment by jerone** [comment]
@DanPristupov Everything appears to be working. Thanks! :+1:

One remark; Azure DevOps only allows only 1 PR per branch ref. So, trying to create an PR on a branch that already has an PR, opens an page that looks like this:
![image](https://user-images.githubusercontent.com/55841/74590811-365e3b00-5012-11ea-9409-1665ee221da5.png)
Fork could use Azure DevOps API to detect if there is already a PR available and open that: https://docs.microsoft.com/en-us/rest/api/azure/devops/git/pull%20requests/get%20pull%20requests?view=azure-devops-rest-4.1#uri-parameters with query-parameter `searchCriteria.sourceRefName`
Example:
```
GET https://dev.azure.com/{organization}/{project}/_apis/git/repositories/{repositoryId}/pullrequests?searchCriteria.sourceRefName=refs/heads/{branchName}&api-version=4.1
```
